runtime/caml/address_class.h

```diff
@@ -15,6 +15,44 @@
 
 /* Classification of addresses for GC and runtime purposes. */
 
+/* The current runtime supports two different configurations that
+   correspond to two different value models, depending on whether
+   "naked pointers" to non-OCaml objects are allowed (considered valid
+   OCaml values).
+
+   In "classic mode", naked pointers are allowed, and the
+   implementation uses a page table. A valid value is then either:
+   - a tagged integer (Is_long or !Is_block from mlvalues.h)
+   - a pointer to the minor heap (Is_young)
+   - a pointer to the major heap (Is_in_heap)
+   - a pointer outside the heap to a well-structured block;
+     an interesting subset of those are the structured constants
+     statically-allocated by OCaml code or the OCaml runtime
+     (Is_in_static_data)
```

Comment:
Nit pick: if the well-structured block is not in a page registered as "static data", i.e. if `Is_in_static_data` returns false, it is treated as a naked pointer.  For example, generic comparison will not traverse it.  So the 4th case could be
"a pointer to a well-structured block, part of a structured constant statically allocated by ocamlopt (Is_in_static_data)".

Reply:
For me, being a naked pointer or not is an extensional property: if you point to a valid OCaml block, you are not naked, naked pointers are those into garbage. (This is the way I understand what some maintainers write when they say that, to avoid naked pointers, one can sometimes use `Caml_out_of_heap_header` for blocks allocated outside the heap.)

This is why I rephrased your message to say that there are well-structured outside-the-heap blocks and naked pointers, and that `Is_in_static_data` is a well-defined *subset* of the well-structured blocks. If a pointer falls outside the "value area", then we don't know whether it is naked or not, and the current implementation must avoid following it (because it may be).

You seem to use a different interpretation of the term "naked pointer", all pointers that go outside the value area. I find it confusing. Are we sure we want to go this way?

(In both interpretations, code pointers are naked pointers.)

Reply:
I rephrased the comment to clarify these nuances in a way that I hope you will be happy with. This part (there are other minor changes) now reads as follows:

```
   - a pointer to a constant block statically-allocated by OCaml code
     or the OCaml runtime (Is_in_static_data)
   - a "foreign" pointer, which is none of the above; the destination
     of those pointers may be a well-formed OCaml blocks, but it may
     also be a naked pointer.

[...] The runtime
conservatively assumes that all foreign pointers may be naked
pointers, and uses the page table to not dereference/follow them.
```